### PR TITLE
Implement cusparse Descriptor class and clean up cusparse code

### PR DESCRIPTION
**Pull request by xwang233**
Add cusparse Descriptor class. Add cusparse Generic API wrapper. Clean up current cuda sparse code.

### CODE REVIEW

**Comment by xwang233** [comment]
cc @csarofeen @ptrblck

File: aten/src/ATen/cuda/CuSparseDescriptors.h [hunk 1]
```diff
@@ -0,0 +1,126 @@
+#pragma once
+
+// guard the whole file
+#if !defined(_MSC_VER) && defined(__CUDACC__) && CUSPARSE_VERSION >= 10301 // CUDA release >= 10.2 and not windows
```
Comment (kurtamohler) [inline aten/src/ATen/cuda/CuSparseDescriptors.h:4]: Shouldn't this be `CUSPARSE_VERSION >= 10200` for v10.2?

Comment (xwang233) [reply aten/src/ATen/cuda/CuSparseDescriptors.h:4]: Hmm, not sure why, but I checked my header file and it's cusparse_version 10301 for my cuda 10.2.89. Also, there was a confirmed mismatch for cusparseSpMM on cuda 10.1

Comment (kurtamohler) [reply aten/src/ATen/cuda/CuSparseDescriptors.h:4]: Oh you're right. I have the same version of cuda and I see 10301 too.

**Review by kurtamohler — APPROVED** [review]
Nice work! BTW, I don't think the `pr/py3.6-clang7-rocmdeb-ubuntu16.04` CI failure is caused by your changes, because I've seen the same thing in a different PR.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**